hub/core/chunk_engine.py

```diff
@@ -0,0 +1,411 @@
+
+        buffer_consumed = self._try_appending_to_last_chunk(buffer, shape)
+        if not buffer_consumed:
+            self._append_to_new_chunk(buffer, shape)
+
+        self.chunk_id_encoder.register_samples_to_last_chunk_id(num_samples)
+
+    def _try_appending_to_last_chunk(
+        self, buffer: memoryview, shape: Tuple[int]
+    ) -> bool:
+        """Will store `buffer` inside of the last chunk if it can.
+        It can be stored in the last chunk if it exists and has space for `buffer`.
+
+        Args:
+            buffer (memoryview): Data to store. This can represent any number of samples.
+            shape (Tuple[int]): Shape for the sample that `buffer` represents.
+
+        Returns:
+            bool: True if `buffer` was successfully written to the last chunk, otherwise False.
+        """
+
+        last_chunk = self.last_chunk
+        if last_chunk is None:
+            return False
+
+        incoming_num_bytes = len(buffer)
+
+        if last_chunk.is_under_min_space(self.min_chunk_size):
+            last_chunk_size = last_chunk.num_data_bytes
+            chunk_ct_content = _min_chunk_ct_for_data_size(
+                self.max_chunk_size, incoming_num_bytes
+            )
+
+            extra_bytes = min(incoming_num_bytes, self.max_chunk_size - last_chunk_size)
+            combined_chunk_ct = _min_chunk_ct_for_data_size(
+                self.max_chunk_size, incoming_num_bytes + last_chunk_size
+            )
+
+            # combine if count is same
+            if combined_chunk_ct == chunk_ct_content:
+                last_chunk.append_sample(
+                    buffer[:extra_bytes], self.max_chunk_size, shape
+                )
+                return True
+
+        return False
+
+    def _append_to_new_chunk(self, buffer: memoryview, shape: Tuple[int]):
+        """Will create a new chunk and store `buffer` inside of it. Assumes that `buffer`'s length is < max chunk size.
+        This should be called if `buffer` could not be added to the last chunk.
+
+        Args:
+            buffer (memoryview): Data to store. This can represent any number of samples.
+            shape (Tuple[int]): Shape for the sample that `buffer` represents.
+        """
+
+        # check if `last_chunk_extended` to handle empty samples
+        new_chunk = self._create_new_chunk()
+        new_chunk.append_sample(buffer, self.max_chunk_size, shape)
+
+    def _create_new_chunk(self):
+        """Creates and returns a new `Chunk`. Automatically creates an ID for it and puts a reference in the cache."""
+
+        chunk_id = self.chunk_id_encoder.generate_chunk_id()
+        chunk = Chunk()
+        chunk_name = ChunkIdEncoder.name_from_id(chunk_id)
+        chunk_key = get_chunk_key(self.key, chunk_name)
+        self.cache[chunk_key] = chunk
+        return chunk
+
+    def extend(self, samples: Union[np.ndarray, Sequence[SampleValue]]):
+        """Formats a batch of `samples` and feeds them into `_append_bytes`."""
```

Comment:
can you simplify this?

Reply:
this code will be simplified once we support large images, and the code isn't straight forward to simplify, so it's ok to leave for later

Reply:
slippery slope